services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java

```diff
@@ -4801,6 +4802,78 @@ public Map<String, StoreDataAudit> getClusterStaleStores(String clusterName) {
+   * @param clusterName, the name of the cluster to search for stores that are ready for compaction
+   * @return a list of StoreInfo objects of stores that are ready for compaction
+   */
+  @Override
+  public ArrayList<StoreInfo> getStoresForCompaction(String clusterName) {
+    ArrayList<StoreInfo> compactionReadyStores = new ArrayList<>();
+    try {
+      Map<String, ControllerClient> childControllers = getVeniceHelixAdmin().getControllerClientMap(clusterName);
+      ArrayList<StoreInfo> storeInfoList = new ArrayList<>();
+
+      // iterate through child controllers
+      for (Map.Entry<String, ControllerClient> controller: childControllers.entrySet()) {
+
+        // add all store info to storeInfoList
+        MultiStoreInfoResponse response = controller.getValue().getClusterStores(clusterName);
+        storeInfoList.addAll(response.getStoreInfoList());
+      }
+
+      // filter out
+      filterStoresForCompaction(storeInfoList, compactionReadyStores);
+    } catch (Exception e) {
+      throw new VeniceException("Something went wrong trying to fetch stores for compaction.", e);
+    }
+    return compactionReadyStores;
+  }
+
+  // package exclusive for testing
+  void filterStoresForCompaction(ArrayList<StoreInfo> storeInfoList, ArrayList<StoreInfo> compactionReadyStores) {
+    for (StoreInfo storeInfo: storeInfoList) {
+      if (isCompactionReady(storeInfo)) {
+        compactionReadyStores.add(storeInfo);
+      }
+    }
+  }
+
+  // This function abstracts the criteria for a store to be ready for compaction
+  boolean isCompactionReady(StoreInfo storeInfo) {
+    boolean isHybridStore = storeInfo.getHybridStoreConfig() != null;
+
+    return isHybridStore && isLastCompactionTimeOlderThanThresholdHours(COMPACTION_THRESHOLD_HOURS, storeInfo);
+  }
+
+  // START isCompactionReady() helper methods: each method below encapsulates a log compaction readiness criterion
+  /**
+   * This function checks if the last compaction time is older than the threshold.
+   * @param compactionThresholdHours, the number of hours that the last compaction time should be older than
+   * @param storeInfo, the store to check the last compaction time for
+   * @return true if the last compaction time is older than the threshold, false otherwise
+   */
+  private boolean isLastCompactionTimeOlderThanThresholdHours(int compactionThresholdHours, StoreInfo storeInfo) {
+    // get the last compaction time
+    int currentVersionNumber = storeInfo.getCurrentVersion();
+    Optional<Version> currentVersion = storeInfo.getVersion(currentVersionNumber);
+    if (!currentVersion.isPresent()) {
+      LOGGER.warn("Couldn't find current version: {} from store: {}", currentVersionNumber, storeInfo.getName());
+      return false; // invalid store because no current version, this store is not eligible for compaction
+    }
+
+    // calculate hours since last compaction
+    long lastCompactionTime = currentVersion.get().getCreatedTime();
```

Comment:
Trying to understand the logic behind this, so we allow compaction where the current version is older than certain threshold ? or should we look at this field `getLatestVersionPromoteToCurrentTimestamp` which tell when the current version was promoted.

Reply:
Good question, Sourav. Bharath and I went through the same consideration. Here's why:

Since the purpose of repush is to compact the store, the time elapsed since version creation is the target metric. Hence, we look at creation rather than promotion timestamp. Does this make more sense?